src/main/java/com/ibm/cloud/sdk/core/util/CredentialUtils.java

```diff
@@ -108,84 +96,96 @@ public static boolean hasBadStartOrEndChar(String credentialValue) {
+    Map<String, List<VcapService>> result = null;
+
+    // Retrieve the environment variable's value.
+    String vcapValue = EnvironmentUtils.getenv(VCAP_SERVICES);
+    if (StringUtils.isNotEmpty(vcapValue)) {
+      Gson gson = GsonSingleton.getGson();
+      // Parse it into a map of VcapService lists keyed by service name.
+      Type typeToken = new TypeToken<Map<String, List<VcapService>>>() { }.getType();
+      try {
+        result = gson.fromJson(vcapValue, typeToken);
+      } catch (Throwable t) {
+        log.log(Level.WARNING, "Error parsing VCAP_SERVICES", t);
+      }
+    }
+    return result;
   }
 
-  /**
-   * Returns the value associated with the provided key from the VCAP_SERVICES, or null if it doesn't exist.
+   /**
+   * Returns an appropriate "service" list item for the specific 'serviceName' value.
    *
-   * @param serviceName the service name
-   * @param key the key whose value should be returned
-   * @param plan the plan name
-   * @return the value of the provided key
+   * @param serviceName the name used to locate the appropriate entry within the VCAP_SERVICES structure.
+   * This value will be used in two ways:
+   *
+   * First, a search will be done to find a service entry whose 'name' field matches the 'vcapName'
+   * parameter value.
+   * Second, if no entry with a matching 'name' field was found, then the first service entry within the
+   * services list whose key matches 'serviceName' will be returned.
+   *
+   *
+   * @return a VcapService instance if found, or null
    */
-  public static String getVcapValue(String serviceName, String key, String plan) {
-    if ((serviceName == null) || serviceName.isEmpty()) {
-      return null;
-    }
-
-    final JsonObject services = getVcapServices();
-    if (services == null) {
+  protected static VcapService getVcapServiceEntry(String serviceName) {
+    // check if param is defined
+    if (serviceName == null || serviceName.isEmpty()) {
       return null;
     }
-
-    JsonObject jsonCredentials = getCredentialsObject(services, serviceName, plan);
-    if (jsonCredentials != null) {
-      if (jsonCredentials.has(key)) {
-        return jsonCredentials.get(key).getAsString();
+    // Retrieve the VCAP_SERVICES environment variable and unmarshal into a Map.
+    Map<String, List<VcapService>> vcapObj = getVcapServicesObj();
+    if (vcapObj != null) {
+      // search the inner entries of the vcap for the matching serviceName
+      for (List<VcapService> serviceList : vcapObj.values()) {
+        for (VcapService service : serviceList) {
+          if (serviceName.equals(service.name)) {
+            return service;
+          }
+        }
+      }
+      // Second, try to find a service list with the specified key.
+      List<VcapService> services = vcapObj.get(serviceName);
+      if (services != null && services.size() > 0) {
+        VcapService service = (services.get(0) != null) ? services.get(0) : null;
```

Comment:
```suggestion
        return services.get(0);
```
this will return either null or a non-null value, which is what we want in both cases

Reply:
Good point. I went ahead and updated that